### PR TITLE
Split okhttp core into Android and JVM

### PR DESCRIPTION
**Pull request by yschimke**
Uses this approach for a shared Java core between Android and JVM.

https://jeroenmols.com/blog/2021/03/17/share-code-kotlin-multiplatform/

Source structure under okhttp

- androidMain
- androidUnitTest
- commonJvmAndroid
- jvmMain
- jvmTest

Outputs

image

TODO

- [x] Get it working fully for running tests
- [x] Get builds working and identify blockers around OSGI, publishing?

### CODE REVIEW

**Comment by yschimke** [comment]
It's noisier because something is persistently picking up the main source set in the android build, and warning about it. Worse it makes builds fail because of duplicate sources.

**Comment by yschimke** [comment]
Also more invasive because a lot is changing here, so building this against 1.9.24 and AGP 8.2 sounds like a dead end.

File: build.gradle.kts [hunk 1]
```diff
@@ -36,16 +42,16 @@ buildscript {
   }
 }
 
-apply(plugin = "org.jetbrains.dokka")
-apply(plugin = "com.diffplug.spotless")
+//apply(plugin = "org.jetbrains.dokka")
+//apply(plugin = "com.diffplug.spotless")
```
Comment (yschimke) [inline build.gradle.kts:46]: Disabled because of Gradle Sync issues when enabled.

File: gradle/libs.versions.toml [hunk 2]
```diff
@@ -5,13 +5,13 @@ com-squareup-moshi = "1.15.1"
 com-squareup-okio = "3.9.1"
 de-mannodermaus-junit5 = "1.6.0"
 graalvm = "22.3.2"
-kotlinx-serialization = "1.6.3"
-ksp = "1.9.24-1.0.20"
+kotlinx-serialization = "1.7.3"
```
Comment (yschimke) [inline gradle/libs.versions.toml:8]: Pretty major version bumps. Seems appropriate for OkHttp 5.

File: okhttp/build.gradle.kts [hunk 3]
```diff
@@ -1,15 +1,16 @@
...
+import com.vanniktech.maven.publish.KotlinMultiplatform
 
 plugins {
-  kotlin("jvm")
+  kotlin("multiplatform")
+  id("com.android.library")
   kotlin("plugin.serialization")
   id("org.jetbrains.dokka")
-  id("com.vanniktech.maven.publish.base")
+//  id("com.vanniktech.maven.publish.base")
```
Comment (swankjesse) [inline okhttp/build.gradle.kts:13]: * [ ] needs something!

File: okhttp-testing-support/src/main/kotlin/okhttp3/internal/concurrent/TaskFaker.kt [hunk 4]
```diff
@@ -360,15 +360,17 @@ class TaskFaker : Closeable {
       timeout: Long,
       unit: TimeUnit,
     ): T? {
-      taskRunner.withLock {
+      return taskRunner.withLock {
```
Comment (swankjesse) [inline okhttp-testing-support/src/main/kotlin/okhttp3/internal/concurrent/TaskFaker.kt:363]: what’s with this change?

File: okcurl/build.gradle.kts [hunk 5]
```diff
@@ -38,9 +38,9 @@ dependencies {
   testImplementation(kotlin("test"))
 }
 
-animalsniffer {
-  isIgnoreFailures = true
-}
+//animalsniffer {
+//  isIgnoreFailures = true
+//}
```
Comment (swankjesse) [inline okcurl/build.gradle.kts:43]: * [ ] needs something

File: gradle/libs.versions.toml (hunk 2 above)
Comment (swankjesse) [reply gradle/libs.versions.toml:8]: Yup

File: build.gradle.kts [hunk 6]
```diff
@@ -204,18 +215,18 @@ subprojects {
...
   }
 
-  tasks.withType<JavaCompile> {
-    sourceCompatibility = JavaVersion.VERSION_1_8.toString()
-    targetCompatibility = JavaVersion.VERSION_1_8.toString()
-  }
+//  tasks.withType<JavaCompile> {
+//    sourceCompatibility = JavaVersion.VERSION_17.toString()
+//    targetCompatibility = JavaVersion.VERSION_17.toString()
+//  }
```
Comment (swankjesse) [inline build.gradle.kts:229]: ??

**Review by swankjesse — APPROVED** [review]
I love it. There’s some Gradle stuff to clean up before landing though!

File: okhttp-testing-support/src/main/kotlin/okhttp3/internal/concurrent/TaskFaker.kt (hunk 4 above)
Comment (yschimke) [reply okhttp-testing-support/src/main/kotlin/okhttp3/internal/concurrent/TaskFaker.kt:363]: Kotlin 2 didn't like the inline with return?

**Comment by yschimke** [comment]
@swankjesse not sure how to avoid plugins that pull in java?

https://github.com/bndtools/bnd/blob/master/gradle-plugins%2Fbiz.aQute.bnd.gradle%2Fsrc%2Fmain%2Fjava%2FaQute%2Fbnd%2Fgradle%2FBndPlugin.java

**Comment by yschimke** [comment]
@swankjesse @JakeWharton heads up, I think it's working now.  Going to land based on prior Jesse approval and work through any issues that pop up.

But I think these are working
- osgi and test (moved to a test module)
- docs and publish to local maven repo
- Android tests
- native image tests

The gradle builds got messier, but I think we should adopt convention plugins, so maybe a good thing to have this in first.

**Comment by yschimke** [comment]
almost!

Working through `android-test-app:minifyReleaseWithR8`

File: build.gradle.kts [hunk 7]
```diff
@@ -211,25 +229,30 @@ subprojects {
...
-    val alpnBootVersion = alpnBootVersion()
-    if (alpnBootVersion != null) {
-      val alpnBootJar = configurations.detachedConfiguration(
-        dependencies.create("org.mortbay.jetty.alpn:alpn-boot:$alpnBootVersion")
-      ).singleFile
-      tasks.withType<Test> {
-        jvmArgs("-Xbootclasspath/p:${alpnBootJar}")
+  if (project.name != "okhttp") {
+    if (platform == "jdk8alpn") {
+      // Add alpn-boot on Java 8 so we can use HTTP/2 without a stable API.
```
Comment (swankjesse) [inline build.gradle.kts:238]: Does this make sense for non-okhttp modules?

File: okhttp/src/commonJvmAndroid/kotlin/okhttp3/internal/connection/Locks.kt [hunk 8]
```diff
@@ -13,7 +13,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-@file:OptIn(ExperimentalContracts::class)
+@file:OptIn(ExperimentalContracts::class, ExperimentalContracts::class)
```
Comment (swankjesse) [inline okhttp/src/commonJvmAndroid/kotlin/okhttp3/internal/connection/Locks.kt:16]: this is an IntelliJ bug

**Review by swankjesse — COMMENTED** [review]
LGTM

Awesome

File: build.gradle.kts (hunk 7 above)
Comment (yschimke) [reply build.gradle.kts:238]: Good call.

I do think our grade files could do with a major redesign/simplification. Maybe less build variants overall.

Each thing like osgi, multiplatform, loom, graal has made them more special case. And extra test runs and platforms to test.

File: okhttp/build.gradle.kts [hunk 9]
```diff
@@ -52,125 +61,192 @@ val generateIdnaMappingTable = tasks.register<JavaExec>("generateIdnaMappingTabl
...
-val copyOsgiTestDeployment = tasks.register<Copy>("copyOsgiTestDeployment") {
-  from(osgiTestDeploy)
-  into(layout.buildDirectory.dir("resources/test/okhttp3/osgi/deployments"))
+    consumerProguardFiles("okhttp3.pro")
+  }
 }
 
-test.configure {
-  dependsOn(copyOsgiTestDeployment)
+// Hack to make BundleTaskExtension pass briefly
```
[the remark this answers is not in the document]
Comment (yschimke) [reply okhttp/build.gradle.kts:187]: I think a later PR links to 1 of 2 bugs I raised for bnd.

Multiplatform support
Lazy configuration support

Currently the plugin assumes that the Java "main" spurceset exists, not jvmMain. 

Removed with a different hack in a later PR

File: build.gradle.kts (hunk 7 above)
Comment (yschimke) [reply build.gradle.kts:238]: https://github.com/square/okhttp/pull/8785

**Pull request merged**